### PR TITLE
Adding support to compare blob content across replicas to BlobValidator

### PR DESCRIPTION
**Pull request by nsivabalan**
Adding support to compare blob content (BlobProperties, UserMetadata and blob data) for replicas of a partition to BlobValidator(--compareContent true)
Adding support to take in a file for list of blobIds to BlobValidator(--blobIdsFilePath [FilePath containing one blobId per line])

SLA: 15 mins
Reviewer: @vgkholla

### CODE REVIEW

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 1]
```diff
@@ -613,6 +664,99 @@ private ServerErrorCode validateBlobOnReplica(BlobId blobId, ClusterMap clusterM
   }
 
   /**
+   * Holds all the data pertaining to a blob like {@link BlobProperties}, UserMetata and the actual blob content
+   */
+  class BlobContent {
+    ServerErrorCode serverErrorCode;
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: all `final`?

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 2]
```diff
@@ -613,6 +664,99 @@ private ServerErrorCode validateBlobOnReplica(BlobId blobId, ClusterMap clusterM
   }
 
   /**
+   * Holds all the data pertaining to a blob like {@link BlobProperties}, UserMetata and the actual blob content
+   */
+  class BlobContent {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: perhaps this is better called `ServerResponse`?

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 3]
```diff
@@ -613,6 +664,99 @@ private ServerErrorCode validateBlobOnReplica(BlobId blobId, ClusterMap clusterM
...
+      if (this.blobProperties.getOwnerId() != null && that.blobProperties.getOwnerId() != null &&
+          (!this.blobProperties.getOwnerId().equals(that.blobProperties.getOwnerId()))) {
+        logger.error("OwnerId Mismatch " + this.blobProperties.getOwnerId() + ", " + that.blobProperties.getOwnerId());
+        return false;
+      } else if (this.blobProperties.getOwnerId() == null || that.blobProperties.getOwnerId() == null) {
+        logger.error("OwnerId Mismatch " + this.blobProperties.getOwnerId() + ", " + that.blobProperties.getOwnerId());
+        return false;
+      }
+
+      if (this.blobProperties.getServiceId() != null && that.blobProperties.getServiceId() != null &&
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: i don't think you the `this` qualifiers are required here.

**Comment by nsivabalan** [comment]
@vgkholla :addressed comments

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 4]
```diff
@@ -297,59 +282,65 @@ public static void main(String args[]) {
...
+        logger.error("Value " + value + " has to be set");
         System.exit(0);
       }
     }
   }
 
-  private ArrayList<BlobId> generateBlobId(ArrayList<String> blobIdListStr, ClusterMap map) throws IOException {
-    ArrayList<BlobId> blobIdList = new ArrayList<BlobId>();
-    for (String blobIdStr : blobIdListStr) {
+  private Set<BlobId> generateBlobId(String blobIdListStr, String blobIdsFilePath, ClusterMap map) throws IOException {
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: `generateBlobIds`?

File: ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java [hunk 5]
```diff
@@ -417,19 +407,41 @@ private void validateBlobOnDatacenter(BlobId blobId, ClusterMap clusterMap, Stri
...
+   */
+  private void compareListOfBlobContent(String blobId, Map<ReplicaId, ServerResponse> replicaIdBlobContentMap) {
+    Iterator<ReplicaId> replicaIdIterator = replicaIdBlobContentMap.keySet().iterator();
+    ReplicaId replicaId1 = replicaIdIterator.next();
+    ServerResponse replica1ServerResponse = replicaIdBlobContentMap.get(replicaId1);
+    while (replicaIdIterator.hasNext()) {
+      ReplicaId replicaId2 = replicaIdIterator.next();
+      ServerResponse replica2ServerResponse = replicaIdBlobContentMap.get(replicaId2);
+      if (!replica1ServerResponse.equals(replica2ServerResponse)) {
+        logger.error(
```
Comment (vgkholla) [inline ambry-tools/src/main/java/com.github.ambry/tools/admin/BlobValidator.java]: please use formats - much easier to read

**Comment by nsivabalan** [comment]
@vgkholla : done

**Pull request merged**